### ISSUE TITLE
"...if any entities in the graph have conflicting key values."

### ISSUE THREAD

**Issue by sam-wheat** [aspnet/EntityFramework6#266]
There are many questions of this nature on Stackoverflow but I can find none that answer my question.

I do not understand why I am getting the "Attaching an entity of type... failed" error message.  As shown in the example code below I am creating a new instance of my context when my entities are retrieved - and another new instance and when they are saved.  

Also, for debugging purposes I am counting modified entities that may exist in the graph immediately before I attach my modified entity.  As expected the counts are zero - but I still get the error.  Why is that?

Entity contains a property of Entity2...
 
```
public class Entity
{
	 public int ID {get; set; }
	 public Entity2 Entity2 {get; set; }
}
```

```
// context is created in a  using block
using(var context = new dbcontext())
{
	Entity e = context.Entities.first(yada);
}
```

```
// context is disposed 
// NEW context created on postback

using(var context = new dbcontext())
{
	// for debugging only:
	var changes = db.ChangeTracker.Entries().ToList();  // Count() = 0!
	var localE = db.Entity.Local.ToList();  // Count() = 0!
	var localE2 = db.Entity2.Local.ToList();  // Count() = 0!
	db.Entity.Attach(e);  // Error here!
}
```

Text of the error message: "...if any entities in the graph have conflicting key values."  
In my code I put a breakpoint on the change counts and observe there are no entities in the graph - are the counts wrong or is the error message misleading?

**Comment by ajcvickers** [comment]
@sam-wheat Can you provide a complete code listing or project that reproduces what you are seeing? There is not enough info in the code posted above to get a good idea of what is going on. For example, when you do `db.Entity.Attach(e);`, what is `db`?

**Comment by sam-wheat** [comment]
@ajcvickers Thank you - I think I solved it:
child is an instance of an entity of type Child.  I was adding a different instance of Child to the Children collection of child's parent object.
So the entity did not already exist in the graph when I started - the problem was that I was trying to add two different objects with the same ID.

```
Child c = new Child { ID = 1, ParentID = 1, Parent = new Parent { ID = 1, Children = new List<Child>() } };
            
// correct:
c.Parent.Children.Add(c);
            
//incorrect:
c.Parent.Children.Add(new Child { ID = 1, ParentID = 1 });

```
Incorrect code above will cause this error:
Attaching an entity of type 'Child' failed because another entity of the same type already has the same primary key value. This can happen when using the 'Attach' method or setting the state of an entity to 'Unchanged' or 'Modified' if any entities in the graph have conflicting key values. This may be because some entities are new and have not yet received database-generated key values. In this case use the 'Add' method or the 'Added' entity state to track the graph and then set the state of non-new entities to 'Unchanged' or 'Modified' as appropriate.
